docs/adr/00xx-moving-business-logic-out-of-controllers.md

```diff
@@ -0,0 +1,124 @@
+
+Date: 2019-06-03
+
+## Context
+
+Over time the controller classes for Content Publisher have become
+[increasingly complex][complex-controllers] as the application grew in size
+and functionality. This has led to action methods that are long, difficult to
+understand and expensive to test. Looking towards the future it can be
+anticipated that there will be more functionality in controllers, such as
+access limiting, concurrent editing protection and invalid state handling.
+Therefore there is an expectation that controllers would become an increasing
+pain point for the application due to their complexity.
+
+This is a common scenario for a growing Rails application. A customary
+approach to address it is to distinguish between business and rendering
+logic in an action - where business logic is the action the user requested,
+such as changing application state, and rendering logic is the process of
+building the response to be returned to the user - and move the business logic
+to another class.
+
+Within the Content Publisher team we have been evaluating
+a number of approaches to perform our business logic operations, such
+as [service objects][],
+[the dry-transaction gem][dry-transaction],
+[the interactor gem][interactor-gem],
+[trailblazer operations][trailblazer-operation] and
+the [Publishing API command pattern][pub-api-command]. We came to the
+following conclusions:
+
+- Service objects: This is a pattern already used in Content Publisher,
+  however it is used to perform distinct supporting tasks rather than being
+  coupled to particular controller actions. We felt the services directory had
+  become a mess in Content Publisher, with little consistency between service
+  objects.
+- Dry-transaction gem: This provides a nice interface for managing response
+  through [ruby blocks][calling-a-dry-transaction]. However dry-transaction
+  requires a strict adherence to a provided DSL which can make a transaction
+  class very different from a plain old Ruby object.
+- Interactor gem: This is a simple, flexible pattern which relies on
+  building up an enhanced open struct based response. It had disadvantages
+  that it wasn't particularly clear what input and output was expected.
+- Trailblazer operation: This gem provides a DSL for creating classes to
+  perform a business operation. It works best within a Rails application
+  that has embraced the collection of Trailblazer utilities.
+- Publishing API command pattern: This is an effective pattern at moving logic
+  out of controllers, it however has produced usage problems where the base
+  class is coupled to particular domain logic.
+
+## Decision
+
+Controller actions can use a dedicated class to perform the business logic of
+a request. Logic in an action is to be focused on producing the appropriate
+HTTP response. To meet this requirement it is expected that classes shall
+exist for actions that mutate application state (typically POST/PUT/DELETE
```

Comment:
This seems more like a consequence or observation, rather than part of the decision. If we happened to have a tiny POST action, would we use an interactor? I thought of this because I was wondering how we could phrase this to get to the interactor bit of the decision earlier, since it helps clarify what the 'classes' are - the suspense was palpable ;-).

Reply:
I've adjusted the section to address this.